src/libraries/System.Net.Http.Json/src/System/Net/Http/Json/TranscodingReadStream.cs

```diff
@@ -0,0 +1,244 @@
+            => throw new NotSupportedException();
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (offset < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(offset));
+            }
+
+            if (count < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            if (buffer.Length - offset < count)
+            {
+                throw new ArgumentOutOfRangeException();
+            }
+
+            var readBuffer = new ArraySegment<byte>(buffer, offset, count);
+            return ReadAsyncCore(readBuffer, cancellationToken);
+        }
+
+        private async Task<int> ReadAsyncCore(ArraySegment<byte> readBuffer, CancellationToken cancellationToken)
+        {
+            if (readBuffer.Count == 0)
+            {
+                return 0;
+            }
+
+            if (_overflowBuffer.Count > 0)
+            {
+                int bytesToCopy = Math.Min(readBuffer.Count, _overflowBuffer.Count);
+                _overflowBuffer.Slice(0, bytesToCopy).CopyTo(readBuffer);
+
+                _overflowBuffer = _overflowBuffer.Slice(bytesToCopy);
+
+                // If we have any overflow bytes, avoid complicating the remainder of the code, by returning as
+                // soon as we copy any content.
+                return bytesToCopy;
+            }
+
+            // Only read more content from the input stream if we have exhausted all the buffered chars.
+            if (_charBuffer.Count == 0)
+            {
+                int bytesRead = await ReadInputChars(cancellationToken).ConfigureAwait(false);
+
+                if (bytesRead == 0)
+                {
+                    // We are done, flush encoder.
+                    _encoder.Convert(_charBuffer.Array!, 0, 0, _byteBuffer.Array!, 0, 0, flush: true, out _, out _, out _);
+                    return 0;
+                }
+            }
+
+            bool completed = false;
+            int charsRead = default;
+            int bytesWritten = default;
+            // If the destination buffer is smaller than GetMaxByteCount(1), we avoid encoding to the destination and we use the overflow buffer instead.
+            if (readBuffer.Count > OverflowBufferSize || _charBuffer.Count == 0)
+            {
+                _encoder.Convert(_charBuffer.Array!, _charBuffer.Offset, _charBuffer.Count, readBuffer.Array!, readBuffer.Offset, readBuffer.Count,
+                    flush: false, out charsRead, out bytesWritten, out completed);
+            }
+
+            _charBuffer = _charBuffer.Slice(charsRead);
+
+            if (completed)
+            {
+                return bytesWritten;
+            }
+            else
```

Comment:
Nit: remove the else.  We're not using cascading if/elses for other cases in this code where the if block returns.